### DISCUSSION TITLE
Collection of filter functions

### DISCUSSION THREAD

**Discussion by e-mair** [sebastienros/jint#1122] [Q&A] [answered]
Hello, I am working on a proof-of-concept application that uses JINT as a form of allowing users to write their own set of filter functions to filter a data set.

The user would write their own JavaScript functions and allow an engine wrapper in C# to pull an array of those functions out from the engine and execute each one in sequence, before returning the filtered data.

My script is as follows:
```
function empNoOver(people) {
    return people.filter(p => {
        return p.EmpNo > 100;
    });
}

function companyIdMustBeOver(people) {
    return companyIdOver(people, 10);
}

const Filters = [
    empNoOver,
    companyIdMustBeOver
];
```

The C# engine wrapper is below:
```
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using Jint.Testing.Core.Models;

namespace JT.Testing.Core.Engine
{
    public class IGPJintEngine
    {
        private Jint.Engine _engine;
        
        public IGPJintEngine()
        {
            _engine = new Jint.Engine(cfg =>
            {
                // Any config goes here.
            });

            var func = new Func<IEnumerable<Person>, int, IEnumerable<Person>>((people, companyId) =>
            {
                return people.Where(p => p.CompanyId > companyId);
            });

            _engine.SetValue("companyIdOver", func);
        }
        
        public IEnumerable<T> Filter<T>(IEnumerable<T> data)
        {
            var script = File.ReadAllText(".\\Scripts\\script.js");
            var functions = _engine.Execute(script).GetValue("Filters");
            
            // TODO: Write the rest of the code.
            
            return new List<T>();
        }
    }
}
```
My problem lies within the Filter<T> method:
```
public IEnumerable<T> Filter<T>(IEnumerable<T> data)
{
     var script = File.ReadAllText(".\\Scripts\\script.js");
     var functions = _engine.Execute(script).GetValue("Filters");
            
     // TODO: Write the rest of the code.
            
     return new List<T>();
}
```
Where the value of "Filters", which is a declared const in my script, always returns null:
![image](https://user-images.githubusercontent.com/88276374/162631418-720ae79f-278c-4a13-a8e0-13cd48eb7ba7.png)

I'm probably being naïve as this seems like a simple thing to do but where am I going wrong? Any help is greatly appreciated!

**Comment by lahma** [comment, marked as answer]
Are using Jint 3.x beta preview from NuGet? `const`/`let` etc will behave very unpredictably with old 2.x version.

**Reply by lahma** [reply]
Ah sorry, the debug window should have told me, IsIntegerIndexArray is 3.x functionality. For reasons I cannot recall from the top of my head, you need to use `engine.Evaluate("Filters")` to get `ArrayInstance` back, `GetValue` goes into more global scope of interop.

**Reply by e-mair** [reply]
We have previously been working with 2.x of Jint, so part of this exercise is figuring out what has changed and what we'll need to change in our codebase with the latest version.

As for my issue: your suggestion to use `engine.Evaluate("Filters")` to get my array of functions works great! Thank you!
